### PR TITLE
Short circuiting invalid context file in context.go

### PR DESCRIPTION
**Pull request by mohammedzee1000**
- Does above
 - Removes old fix for component list as this covers that
   more easily
 - Does some breakup of functions of context.go along the way.

Signed-off-by: Mohammed Zeeshan Ahmed <mohammed.zee1000@gmail.com>

## Was the change discussed in an issue?
#1486 

## How to test changes?

Go into a directory where there is no .odo file and try `odo app list` or `odo component list`

### CODE REVIEW

File: pkg/odo/genericclioptions/context.go [hunk 1]
```diff
@@ -72,6 +72,42 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
 	}
 }
 
+func checkValidConfig(command *cobra.Command, configFileName string) {
```
Comment (girishramnani) [inline pkg/odo/genericclioptions/context.go]: error should be returned

File: pkg/odo/genericclioptions/context.go [hunk 2]
```diff
@@ -72,6 +72,42 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
...
+	allowedFirstChildCommandList := []string{"project", "catalog", "login", "logout"}
+
+	if command.HasParent() {
+		rootCommand := command.Root()
+		// if command is create operation or push operation just allow it
+		if command.Name() == "create" || command.Name() == "push" {
+			return
+		}
+		// Find the first child of the command. At end of this, c will be the first child
+		p := command.Parent()
```
Comment (girishramnani) [inline pkg/odo/genericclioptions/context.go]: extracting this out in a function like maybe `getFirstChildOfCommand()` will make this more understandable. just a suggestion

File: pkg/odo/genericclioptions/context.go [hunk 3]
```diff
@@ -72,6 +72,42 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
 	}
 }
 
+func checkValidConfig(command *cobra.Command, configFileName string) {
+	_, err1 := config.NewLocalConfigInfo(configFileName, true)
+	allowedFirstChildCommandList := []string{"project", "catalog", "login", "logout"}
```
Comment (girishramnani) [inline pkg/odo/genericclioptions/context.go]: maybe add a comment which explains that these commands dont need config and hence can run from directories which do not have a `.odo`

File: pkg/odo/genericclioptions/context.go [hunk 4]
```diff
@@ -72,6 +72,42 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
 	}
 }
 
+func checkValidConfig(command *cobra.Command, configFileName string) {
+	_, err1 := config.NewLocalConfigInfo(configFileName, true)
```
Comment (girishramnani) [inline pkg/odo/genericclioptions/context.go]: why call this `err1`?

File: pkg/odo/genericclioptions/context.go [hunk 5]
```diff
@@ -147,6 +169,46 @@ func newContext(command *cobra.Command, createAppIfNeeded bool) *Context {
...
 		}
 	}
+	return app
+}
+
+// resolveComponent resolves component
+func resolveComponent(command *cobra.Command, configFileName string, context *Context) string {
+	var cmp string
+	lci, err := config.NewLocalConfigInfo(configFileName, false)
+	util.LogErrorAndExit(err, "could not get component settings from config file")
```
Comment (girishramnani) [inline pkg/odo/genericclioptions/context.go]: I would like all the logging to be doing by the `GenericRun` and these function can just return error. just a suggestion. maybe @kadel can confirm this

Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Agreed. But that will require a bigger refactor as this is pattern in most of the file. Maybe an issue to track it

File: pkg/odo/genericclioptions/context.go (hunk 2 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Done

File: pkg/odo/genericclioptions/context.go (hunk 3 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Added

**Comment by mohammedzee1000** [comment]
Will add more tests once people are ok with flow

File: pkg/odo/genericclioptions/context.go [hunk 6]
```diff
@@ -147,6 +186,46 @@ func newContext(command *cobra.Command, createAppIfNeeded bool) *Context {
...
+	}
+	return cmp
+}
+
+// newContext creates a new context based on the command flags, creating missing app when requested
+func newContext(command *cobra.Command, createAppIfNeeded bool) *Context {
+	client := client(command)
+
+	// resolve project
+	ns := resolveProject(command, client)
```
Comment (anmolbabu) [inline pkg/odo/genericclioptions/context.go]: Can this be done below after the config resolution?

File: pkg/odo/genericclioptions/context.go [hunk 7]
```diff
@@ -72,6 +72,59 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
...
+			c = p
+			p = c.Parent()
+		}
+	}
+	return nil
+}
+
+func checkValidConfig(command *cobra.Command, configFileName string) {
+	_, err := config.NewLocalConfigInfo(configFileName, true)
+	// List of allowed firstchild command
```
Comment (anmolbabu) [inline pkg/odo/genericclioptions/context.go]: err ignored?

File: pkg/odo/genericclioptions/context.go [hunk 8]
```diff
@@ -72,6 +72,59 @@ func checkProjectCreateOrDeleteOnlyOnInvalidNamespace(command *cobra.Command, er
...
+			// Traverse backwards making current command as the parent and parent as the grandparent
+			c = p
+			p = c.Parent()
+		}
+	}
+	return nil
+}
+
+func checkValidConfig(command *cobra.Command, configFileName string) {
+	_, err := config.NewLocalConfigInfo(configFileName, true)
```
Comment (anmolbabu) [inline pkg/odo/genericclioptions/context.go]: NewLocalConfigInfo actually reads from file everytime its invoked
I believe the aim of this PR is also to decrease the number of times its called which is the reason I believe why we are moving config resolution to context
In my opinion component settings is parallel to context except for only one field in context which is client so should Context struct also be made to contain LocalConfigInfo? and remove project, component name and app from context bcoz they are there in LocalConfigInfo

File: pkg/odo/genericclioptions/context.go (hunk 7 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Not really. It gets caught as the config is read again later with 2nd parameter set to false. This is specifically meant to fail if file is not found

File: pkg/odo/genericclioptions/context.go (hunk 8 above)
Comment (anmolbabu) [reply pkg/odo/genericclioptions/context.go]: in which case there probably is no need to resolve each of project, component and app in my opinion bcoz now the single source of truth in my opinion should be config

Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Unless overridden by respective parameters :)

Comment (anmolbabu) [reply pkg/odo/genericclioptions/context.go]: https://github.com/openshift/odo/issues/1520#issuecomment-477943538

No flags related to component or app needs to be passed to any command other than component create, if passed they'll be ignored or will error out

Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Cool then

File: pkg/odo/genericclioptions/context.go (hunk 6 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: done

File: pkg/odo/genericclioptions/context.go (hunk 7 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Handled

File: pkg/odo/genericclioptions/context.go (hunk 8 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Updated. i really feel like removing those flags right now. Already have it in my git stash :)

File: pkg/odo/genericclioptions/context.go (hunk 1 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Did the best i could for now, it is returned and handled by caller. But will require bigger refactor

**Comment by mohammedzee1000** [comment]
@cdrage it seems to be like it already is as far as context.go is concerned. What is important is shifting to making use of it

**Comment by mohammedzee1000** [comment]
/cc @anmolbabu @girishramnani 
Anything else?

File: pkg/odo/genericclioptions/context.go [hunk 9]
```diff
@@ -147,6 +211,47 @@ func newContext(command *cobra.Command, createAppIfNeeded bool) *Context {
...
+	lci, err := getValidConfig(command)
+	if err != nil {
+		util.LogErrorAndExit(err, "")
+	}
+
+	// resolve project
+	ns := resolveProject(command, client, lci)
+
+	// resolve application
+	app := resolveApp(command, createAppIfNeeded, lci)
```
Comment (anmolbabu) [inline pkg/odo/genericclioptions/context.go]: I don't see the importance of this bcoz it's only going to be component create that will not pass these flagsas far as I can think of but probably defer it to post new workflow implementation for all commands after which we can probably revisit this

**Comment by mohammedzee1000** [comment]
> Could we have e2e tests with the new framework? Suggestions @kadel

I am not sure if we want to do this right now. Lets setup separate prs to move all the tests over properly

File: pkg/odo/genericclioptions/context.go (hunk 9 above)
Comment (mohammedzee1000) [reply pkg/odo/genericclioptions/context.go]: Yes we can revisit this after. Its more about making things clear and concise. It also makes this function much smaller which is what is favoured, especially in go

**Comment by amitkrout** [comment]
> > Could we have e2e tests with the new framework? Suggestions @kadel
> 
> I am not sure if we want to do this right now. Lets setup separate prs to move all the tests over properly

Anyway we have stable e2e test and only effort will be migrate them to new structure. So separate pr would be fine.

**Comment by mohammedzee1000** [comment]
Interesting fail
/retest

**Comment by mohammedzee1000** [comment]
Cluster problem
/retest

File: pkg/config/config.go [hunk 10]
```diff
@@ -77,8 +77,9 @@ type proxyLocalConfig struct {
 // LocalConfigInfo wraps the local config and provides helpers to
 // serialize it.
 type LocalConfigInfo struct {
-	Filename    string `yaml:"FileName,omitempty"`
-	LocalConfig `yaml:",omitempty"`
+	Filename         string `yaml:"FileName,omitempty"`
+	LocalConfig      `yaml:",omitempty"`
+	ConfigFileExists bool
```
Comment (anmolbabu) [inline pkg/config/config.go]: Needs to be an un-exported field exposed if required only be a getter and no setter needs to be there
as this ideally needs to be set only by New... atleast for now

Comment (mohammedzee1000) [reply pkg/config/config.go]: Makes sense, updating

**Comment by mohammedzee1000** [comment]
there was an actual issue in the test
/cc @girishramnani

**Pull request merged**